File: practical-06-mono-polyalphabetic-cipher/poly_alphabetic_cipher.py
import string


def create_monoalphabetic_key(key_word):
    """Create a monoalphabetic substitution key from a keyword."""
    key_word = key_word.upper().replace(" ", "")
    seen = []
    for char in key_word:
        if char not in seen:
            seen.append(char)

    alphabet = string.ascii_uppercase
    remaining = [c for c in alphabet if c not in seen]
    return seen + remaining


def monoalphabetic_encrypt(plaintext, key_word):
    """Encrypt using monoalphabetic substitution with keyword-based key."""
    key_map = create_monoalphabetic_key(key_word)
    normal = string.ascii_uppercase

    result = ""
    for char in plaintext:
        if char.isupper():
            idx = normal.index(char)
            result += key_map[idx]
        elif char.islower():
            idx = normal.index(char.upper())
            result += key_map[idx].lower()
        else:
            result += char
    return result


def monoalphabetic_decrypt(ciphertext, key_word):
    """Decrypt using monoalphabetic substitution with keyword-based key."""
    key_map = create_monoalphabetic_key(key_word)
    normal = string.ascii_uppercase

    result = ""
    for char in ciphertext:
        if char.isupper():
            idx = key_map.index(char)
            result += normal[idx]
        elif char.islower():
            idx = key_map.index(char.upper())
            result += normal[idx].lower()
        else:
            result += char
    return result


def vigenere_encrypt(plaintext, keyword):
    """Encrypt using Vigenère cipher (polyalphabetic substitution)."""
    keyword = keyword.upper().replace(" ", "")
    result = ""
    key_index = 0

    for char in plaintext:
        if char.isalpha():
            shift = ord(keyword[key_index % len(keyword)]) - 65
            if char.isupper():
                result += chr((ord(char) - 65 + shift) % 26 + 65)
            else:
                result += chr((ord(char) - 97 + shift) % 26 + 97)
            key_index += 1
        else:
            result += char
    return result


def vigenere_decrypt(ciphertext, keyword):
    """Decrypt using Vigenère cipher (polyalphabetic substitution)."""
    keyword = keyword.upper().replace(" ", "")
    result = ""
    key_index = 0

    for char in ciphertext:
        if char.isalpha():
            shift = ord(keyword[key_index % len(keyword)]) - 65
            if char.isupper():
                result += chr((ord(char) - 65 - shift) % 26 + 65)
            else:
                result += chr((ord(char) - 97 - shift) % 26 + 97)
            key_index += 1
        else:
            result += char
    return result


if __name__ == "__main__":
    text = "Hello World"
    mono_key = "CRYPTO"
    poly_key = "KEY"

    print("=== Monoalphabetic Cipher ===")
    encrypted_mono = monoalphabetic_encrypt(text, mono_key)
    decrypted_mono = monoalphabetic_decrypt(encrypted_mono, mono_key)
    print(f"Original: {text}")
    print(f"Key word: {mono_key}")
    print(f"Encrypted: {encrypted_mono}")
    print(f"Decrypted: {decrypted_mono}")

    print("\n=== Polyalphabetic Cipher (Vigenère) ===")
    encrypted_poly = vigenere_encrypt(text, poly_key)
    decrypted_poly = vigenere_decrypt(encrypted_poly, poly_key)
    print(f"Original: {text}")
    print(f"Keyword: {poly_key}")
    print(f"Encrypted: {encrypted_poly}")
    print(f"Decrypted: {decrypted_poly}")

File: c programs/poly_alphabetic_cipher.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

void create_monoalphabetic_key(const char *key_word, char key_map[]) {
    int seen[26] = {0};
    int key_idx = 0;
    
    for (int i = 0; key_word[i] != '\0'; i++) {
        char c = toupper(key_word[i]);
        if (c >= 'A' && c <= 'Z' && !seen[c - 'A']) {
            seen[c - 'A'] = 1;
            key_map[key_idx++] = c;
        }
    }
    
    for (char c = 'A'; c <= 'Z'; c++) {
        if (!seen[c - 'A']) {
            key_map[key_idx++] = c;
        }
    }
}

void monoalphabetic_encrypt(const char *plaintext, const char *key_word, char *result) {
    char key_map[26];
    create_monoalphabetic_key(key_word, key_map);
    
    int i = 0;
    for (i = 0; plaintext[i] != '\0'; i++) {
        if (isupper(plaintext[i])) {
            result[i] = key_map[plaintext[i] - 'A'];
        } else if (islower(plaintext[i])) {
            result[i] = tolower(key_map[toupper(plaintext[i]) - 'A']);
        } else {
            result[i] = plaintext[i];
        }
    }
    result[i] = '\0';
}

void monoalphabetic_decrypt(const char *ciphertext, const char *key_word, char *result) {
    char key_map[26];
    create_monoalphabetic_key(key_word, key_map);
    
    int i = 0;
    for (i = 0; ciphertext[i] != '\0'; i++) {
        if (isupper(ciphertext[i])) {
            for (int j = 0; j < 26; j++) {
                if (key_map[j] == ciphertext[i]) {
                    result[i] = 'A' + j;
                    break;
                }
            }
        } else if (islower(ciphertext[i])) {
            for (int j = 0; j < 26; j++) {
                if (key_map[j] == toupper(ciphertext[i])) {
                    result[i] = tolower('A' + j);
                    break;
                }
            }
        } else {
            result[i] = ciphertext[i];
        }
    }
    result[i] = '\0';
}

void vigenere_encrypt(const char *plaintext, const char *keyword, char *result) {
    int key_len = strlen(keyword);
    int key_index = 0;
    
    for (int i = 0; plaintext[i] != '\0'; i++) {
        if (isalpha(plaintext[i])) {
            int shift = toupper(keyword[key_index % key_len]) - 'A';
            if (isupper(plaintext[i])) {
                result[i] = (char)((plaintext[i] - 'A' + shift) % 26 + 'A');
            } else {
                result[i] = (char)((plaintext[i] - 'a' + shift) % 26 + 'a');
            }
            key_index++;
        } else {
            result[i] = plaintext[i];
        }
    }
    result[ strlen(plaintext) ] = '\0';
}

void vigenere_decrypt(const char *ciphertext, const char *keyword, char *result) {
    int key_len = strlen(keyword);
    int key_index = 0;
    
    for (int i = 0; ciphertext[i] != '\0'; i++) {
        if (isalpha(ciphertext[i])) {
            int shift = toupper(keyword[key_index % key_len]) - 'A';
            if (isupper(ciphertext[i])) {
                result[i] = (char)((ciphertext[i] - 'A' - shift + 26) % 26 + 'A');
            } else {
                result[i] = (char)((ciphertext[i] - 'a' - shift + 26) % 26 + 'a');
            }
            key_index++;
        } else {
            result[i] = ciphertext[i];
        }
    }
    result[ strlen(ciphertext) ] = '\0';
}

int main() {
    const char *text = "Hello World";
    const char *mono_key = "CRYPTO";
    const char *poly_key = "KEY";
    
    char encrypted_mono[100], decrypted_mono[100];
    char encrypted_poly[100], decrypted_poly[100];
    
    printf("=== Monoalphabetic Cipher ===\n");
    monoalphabetic_encrypt(text, mono_key, encrypted_mono);
    monoalphabetic_decrypt(encrypted_mono, mono_key, decrypted_mono);
    printf("Original: %s\n", text);
    printf("Key word: %s\n", mono_key);
    printf("Encrypted: %s\n", encrypted_mono);
    printf("Decrypted: %s\n", decrypted_mono);
    
    printf("\n=== Polyalphabetic Cipher (Vigenère) ===\n");
    vigenere_encrypt(text, poly_key, encrypted_poly);
    vigenere_decrypt(encrypted_poly, poly_key, decrypted_poly);
    printf("Original: %s\n", text);
    printf("Keyword: %s\n", poly_key);
    printf("Encrypted: %s\n", encrypted_poly);
    printf("Decrypted: %s\n", decrypted_poly);
    
    return 0;
}
